docs/proposal-hosting-option.md

```diff
@@ -0,0 +1,75 @@
+
+## Why
+
+By building a centralised (but still self-hostable) hosting platform that could connect to anybody's Foam repo, we get the following:
+
+* A zero-effort place for less-technical users to have their notes published
+* A centralised place for a "standard" set of non-standard React components to be defined/reviewed/published and then accessible with MDX
+* A separate collaboration around themes/design/custom components for displaying notes on the web
+
+As long as the code is capable of dynamically fetching & rendering _any_ foam repo, there's actually nothing special about the central copy: anyone can host their own Foam renderer under their own URL, either locked to their own content or open for others to preview with their own data.
+
+By separating the publication ideas from the content ideas (and making them interoperable) I think we'd end up with much more cool ideas/experimentation than tying them together.
+
+## Ideal architecture
+
+* A user on GitHub (`<username>`) creates a public Foam repo name `foam-notes` (or something standardised)
+* The URL `<username>.foam.website` (or something similar) is immediately available as a public view of their notes.
+* Behind the scenes, by visiting a new username, the metadata for building up backlinks etc is built up in the background.
+* Once the site has been crawled & cached, performance should be indistinguishable from self-hosted static files.
+* Content changes are propagated within seconds with clever background fetching & caching
+
+For non-GitHub users, it would be easy to set up the same sort of behaviour under a URL scheme `<gitlab-username>.gitlab.foam.website`.
+
+## How
+
+I could give a technical pitch for why I want to build it this way, but the truth is: I want to build it as a tech demo/showcase of what [FABs](https://fab.dev) can do running on an edge, using [Cloudflare Workers](https://workers.dev), deployed using [Linc](https://linc.sh). So, my proposal:
+
+* NextJS in server-rendered mode
+* Hosted in a FAB in Cloudflare Workers
+* Using the [KV store](https://developers.cloudflare.com/workers/reference/storage) as a global cache for metadata/rendered documents.
+
+For the first request for any new URL:
+
+* Fetch the raw markdown from GitHub directly
+* Store it in the Cache
+* Return rendered page
+
+For repo-wide metadata like backlinks, we don't need to slow down the first request, but can set off a crawler process in the background.
+
+Both of these processes lend themselves to a "serve stale" approach:
+
+* The edge renderer _always_ responds with the current version in-cache, so responses are always fast (after the very first requests).
+* Periodically (ideally, once every ~5-10 seconds) GitHub is checked to see if there have been updates. If there have been, the cache is refreshed and new requests will get the updated content. **This means the time between pushing new content and it being visible could be as low as 5 seconds**.
```

Comment:
There could be an hybrid for "advanced" users where you ask them to add a webhook on their repo which ping the service when they push. That way you don't have to poll all of the repos and GH is happier.

Reply:
Yep definitely, I have a caching model in mind for the polling that would make it pretty minimal on GH, but you could definitely move to a push model. I still think the idea that there's zero configuration required on the author's part (and you still get sub-10-second caching) is the most interesting thing though.

It also means that it's not just Foam that this would work for, but any crawlable collection of Markdown files...

Reply:
>  I still think the idea that there's zero configuration required on the author's part is the most interesting thing though

Yep agreed. I was more thinking about the next step if someone wants even more real time updates